lib/clients/GcpClient.py

```diff
@@ -0,0 +1,553 @@
+
+        self.__gcpCredentials = json.loads(configuration['credentials'])
+        self.project_id = configuration['projectId']
+        self.tags = {
+            'instance_id': self.INSTANCE_ID,
+            'job_name': self.JOB_NAME
+        }
+
+        self.compute_api_name = 'compute'
+        self.compute_api_version = 'v1'
+        self.snapshot_prefix = 'sf-snapshot'
+        self.disk_prefix = 'sf-disk'
+        self.device_path_template = '/dev/disk/by-id/google-{}'
+
+        self.compute_client = self.create_compute_client()
+        self.storage_client = self.create_storage_client()
+
+        # +-> Check whether the given container exists and is accessible
+        self.container = self.get_container()
+        if not self.container:
+            msg = 'Could not find or access the given container.'
+            self.last_operation(msg, 'failed')
+            raise Exception(msg)
+
+        # +-> Get the availability zone of the instance
+        self.availability_zone = self._get_availability_zone_of_server(
+            configuration['instance_id'])
+        if not self.availability_zone:
+            msg = 'Could not retrieve the availability zone of the instance.'
+            self.last_operation(msg, 'failed')
+            raise Exception(msg)
+
+    def create_compute_client(self):
+        try:
+            credentials = service_account.Credentials.from_service_account_info(
+                self.__gcpCredentials)
+            compute_client = discovery.build(
+                self.compute_api_name, self.compute_api_version, credentials=credentials)
+            return compute_client
+        except Exception as error:
+            raise Exception(
+                'Creation of compute client failed: {}'.format(error))
+
+    def create_storage_client(self):
+        try:
+            credentials = service_account.Credentials.from_service_account_info(
+                self.__gcpCredentials)
+            storage_client = storage.Client(self.project_id, credentials)
+            return storage_client
+        except Exception as error:
+            raise Exception(
+                'Creation of storage client failed: {}'.format(error))
+
+    def _get_availability_zone_of_server(self, instance_id):
+        try:
+            expression = 'name eq ' + instance_id
+            request = self.compute_client.instances().aggregatedList(
+                project=self.project_id, filter=expression)
+            while request is not None:
+                response = request.execute()
+                for zone, instances_scoped_list in response['items'].items():
+                    if 'instances' in instances_scoped_list:
+                        for instance in instances_scoped_list['instances']:
+                            if instance['name'] == instance_id:
+                                # Assuming the last part of the url is zone-name
+                                return instance['zone'].rsplit('/', 1)[1]
```

Comment:
Out of curiosity, how does the zone attribute look like? Wondering if there are exceptions to the rule!

Reply:
Example zone attribute looks like:
``` "zone": "https://www.googleapis.com/compute/v1/projects/project_name/zones/europe-west1-b" ```
Specification of the api used here is at: [instances/aggregatedList](https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList). Here **"instances"** details under "**key**" refer to [instances#resource](https://cloud.google.com/compute/docs/reference/latest/instances#resource). In this, zone refers to 'URL of the zone where the instance resides.' and this URL specification can be found in [zones/get](https://cloud.google.com/compute/docs/reference/latest/zones/get).